### PR TITLE
Changeup gate check

### PR DESCRIPTION
**Pull request by tadgh**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java [hunk 1]
```diff
@@ -270,6 +269,14 @@ public void incrementWorkChunkErrorCount(String theChunkId, int theIncrementBy)
 		myWorkChunkRepository.incrementWorkChunkErrorCount(theChunkId, theIncrementBy);
 	}
 
+	@Override
+	@Transactional(propagation = Propagation.REQUIRES_NEW)
+	public boolean canAdvanceInstanceToNextStep(String theInstanceId, String theCurrentStepId) {
+		List<StatusEnum> statusesForStep = myWorkChunkRepository.getDistinctStatusesForStep(theInstanceId, theCurrentStepId);
+		ourLog.debug("Checking whether gated job can advanced to next step. [instanceId={}, currentStepId={}, statusesForStep={}]", theInstanceId, theCurrentStepId, statusesForStep);
+		return statusesForStep.stream().noneMatch(StatusEnum::isIncomplete);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:277]: Maybe check it's not empty as well?

Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:277]: Or better yet that it includes COMPLETE

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/SynchronizedJobPersistenceWrapper.java [hunk 2]
```diff
@@ -166,4 +171,9 @@ public synchronized boolean markInstanceAsCompleted(String theInstanceId) {
 	public JobOperationResultJson cancelInstance(String theInstanceId) {
 		return myWrap.cancelInstance(theInstanceId);
 	}
+
+	@Override
+	public List<String> fetchAllChunkIdsForStep(String theInstanceId, String theNextStepId) {
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/coordinator/SynchronizedJobPersistenceWrapper.java:176]: should be theStepId

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/api/IJobPersistence.java [hunk 3]
```diff
@@ -226,4 +234,6 @@ default Optional<WorkChunk> markWorkChunkAsErroredAndIncrementErrorCount(MarkWor
 	 * @param theInstanceId The instance ID
 	 */
 	JobOperationResultJson cancelInstance(String theInstanceId);
+
+	List<String> fetchAllChunkIdsForStep(String theInstanceId, String theNextStepId);
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/api/IJobPersistence.java:238]: should be theStepId

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java [hunk 4]
```diff
@@ -156,18 +152,23 @@ private void triggerGatedExecutions() {
...
 		}
 	}
 
 	private void processChunksForNextSteps(String instanceId, String nextStepId) {
-		List<String> chunksForNextStep = myProgressAccumulator.getChunkIdsWithStatus(instanceId, nextStepId, StatusEnum.QUEUED);
-		for (String nextChunkId : chunksForNextStep) {
+		List<String> queuedChunksForNextStep = myProgressAccumulator.getChunkIdsWithStatus(instanceId, nextStepId, StatusEnum.QUEUED);
+		int totalChunksForNextStep = myProgressAccumulator.getTotalChunkCountForInstanceAndStep(instanceId, nextStepId);
+		if (totalChunksForNextStep != queuedChunksForNextStep.size()) {
+			ourLog.error("Total chunk size to submit for next step does not match QUEUED chunk size! [instanceId={}, stepId={}, totalChunks={}, queuedChunks={}]", instanceId, nextStepId, totalChunksForNextStep, queuedChunksForNextStep.size());
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java:164]:
```suggestion
			ourLog.warn("Total progress accumulator QUEUED chunk count does not match database QUEUED chunk count! [instanceId={}, stepId={}, totalChunks={}, queuedChunks={}]", instanceId, nextStepId, totalChunksForNextStep, queuedChunksForNextStep.size());
```

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java [hunk 5]
```diff
@@ -156,18 +152,23 @@ private void triggerGatedExecutions() {
...
+		}
+		List<String> chunksToSubmit = myJobPersistence.fetchAllChunkIdsForStep(instanceId, nextStepId);
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java:166]: What if some of these chunks are not in QUEUED state.  Technically this shouldn't happen, but we don't guarantee it here.  I guess maybe do we catch it down stream in the work processor if it picks up a job that is not in QUEUED state?

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java (hunk 5 above)
Comment (tadgh) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/maintenance/JobInstanceProcessor.java:166]: FWIW, chunks are created in queued state, and gated execution prevents any processing on them before the point at which we submit them. I think we're OK here.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java (hunk 1 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:277]: Sure thing

**Pull request merged**